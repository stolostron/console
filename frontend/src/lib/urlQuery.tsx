/* Copyright Contributors to the Open Cluster Management project */

// This function updates url query string based on page filter selections
// export function updateBrowserUrlFilterPresets(filters, search, sort) {
//     // TODO
// }

/**
 * transformBrowserUrlToFilterPresets returns preset filters, search text, sorting, current page number and items per page defined in the window.location.search.
 *
 * initialFilters - An object of { [key: string]: string[] } entries
 * Where key matches the filter category id (should be a lowercase string)
 * Where string[] contains strings of the selected category option values
 *
 * initialSearch - A string to use in the table searchbar.
 *
 * TODO update to use the column name instead of the index to support column management
 * initialSort - Index of the column to sort on, prefixed with an '-' or '' to denote ascending or descending sorting.
 *
 * initialPage - the current page number.
 *
 * initialPerPage - How many items to display per page.
 *
 * Example usage: ?region=us-east&region=us-south&status=online&status=offline&search=some text goes here&sort=-0&page=5&perPage=10
 * Example response: {"initialFilters":{"id":"region","options":["us-east","us-south"]},{"id":"status","options":["online","offline"]},"initialSearch":"some text goes here","initialSort":{"index":0,"direction":"asc"},"initialPage":5,"initialPerPage":10}
 * */
export function transformBrowserUrlToFilterPresets(urlSearch: string) {
  const initialFilters: { [key: string]: string[] } = {}
  let initialSearch = ''
  let initialSort: { index: number; direction: 'asc' | 'desc' } = { index: 0, direction: 'asc' }
  let initialPage = 1
  let initialPerPage = 10
  const urlparams = new URLSearchParams(decodeURIComponent(urlSearch))
  const entries = urlparams.entries()
  let entry = entries.next()
  while (!entry.done) {
    const key = entry.value[0]
    const value = entry.value[1]
    switch (key) {
      case 'search':
        initialSearch = value
        break
      case 'sort':
        initialSort = {
          index: Number.isInteger(Number(value)) ? Math.abs(Number(value)) : 0,
          direction: value.startsWith('-') ? 'asc' : 'desc',
        }
        break
      case 'page':
        if (Number.isInteger(Number(value))) {
          initialPage = Number(value)
        }
        break
      case 'perPage':
        if (Number.isInteger(Number(value))) {
          initialPerPage = Number(value)
        }
        break
      default:
        // If the key doesn't match one of the above we assume it is a filter key
        if (!initialFilters[key]) {
          initialFilters[key] = [value]
        } else {
          initialFilters[key].push(value)
        }
    }
    entry = entries.next()
  }

  return { initialFilters, initialSearch, initialSort, initialPage, initialPerPage }
}
