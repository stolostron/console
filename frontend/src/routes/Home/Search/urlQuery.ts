/* Copyright Contributors to the Open Cluster Management project */
// Copyright (c) 2021 Red Hat, Inc.
// Copyright Contributors to the Open Cluster Management project
import queryString from 'query-string'

// This function pushes the new search query to the browsers history
export function updateBrowserUrl(history: any, currentQuery: string) {
  if (currentQuery === '') {
    // on clear search query
    history.push(window.location.pathname)
  } else {
    const url = `${window.location.pathname}?filters={"textsearch":${encodeURIComponent(JSON.stringify(currentQuery))}}`
    history.push(url)
  }
}

// This function handles navigation to search page with a predefined searchquery in browsers url
export function transformBrowserUrlToSearchString(urlQuery: string) {
  // Example browser url string:
  // .../multicloud/home/search?filters={"textsearch":"kind%3Adeployment%20name%3Asearch-prod-df8fa-search-api"}&showrelated=pod
  const presetSearchQuery = ''
  const preSelectedRelatedResources: string[] = []
  if (urlQuery !== '') {
    const paramString = queryString.parse(urlQuery)

    //Filter out the search query string
    let filterString = paramString.filters?.toString()
    // cut the start of the url string
    filterString = filterString?.replace('{"textsearch":"', '')
    // cut the end of the url string
    filterString = filterString?.replace('"}', '')

    // get any pre-selected related resources
    const relatedString = paramString.showrelated?.toString().toLowerCase()
    const relatedArray = relatedString?.split(',')

    return {
      presetSearchQuery: filterString,
      preSelectedRelatedResources: relatedArray,
    }
  }
  return { presetSearchQuery, preSelectedRelatedResources }
}
