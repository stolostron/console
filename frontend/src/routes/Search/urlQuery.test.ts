/* Copyright Contributors to the Open Cluster Management project */
// Copyright (c) 2021 Red Hat, Inc.
// Copyright Contributors to the Open Cluster Management project
import { transformBrowserUrlToSearchString, updateBrowserUrl } from './urlQuery'

test('Correctly returns transformBrowserUrlToSearchString', () => {
  const testString = 'filters={"textsearch":"kind%3Adeployment%20name%3Asearch-prod-df8fa-search-api"}&showrelated=pod'
  const result = transformBrowserUrlToSearchString(testString)
  expect(result).toMatchSnapshot()
})

test('Correctly returns updateBrowserUrl', () => {
  const navigate = jest.fn()
  const testData = 'kind:deployment name:search-prod-df8fa-search-api kind:'
  updateBrowserUrl(navigate, testData)
  expect(navigate).toHaveBeenCalled()
})

test('Correctly returns updateBrowserUrl with empty query string', () => {
  const navigate = jest.fn()
  updateBrowserUrl(navigate, '')
  expect(navigate).toHaveBeenCalled()
})
