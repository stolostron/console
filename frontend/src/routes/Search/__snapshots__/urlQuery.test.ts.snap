// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Correctly returns transformBrowserUrlToSearchString 1`] = `
Object {
  "preSelectedRelatedResources": Array [
    "pod",
  ],
  "presetSearchQuery": "kind:deployment name:search-prod-df8fa-search-api",
}
`;
